controllers/virtualmachinepublishrequest/virtualmachinepublishrequest_controller.go

```diff
@@ -511,6 +522,53 @@ func (r *Reconciler) getItemByName(ctx *pkgctx.VirtualMachinePublishRequestConte
 	return r.VMProvider.GetItemFromLibraryByName(ctx, contentLibrary, itemName)
 }
 
+func (r *Reconciler) checkContentLibraryQuota(ctx *pkgctx.VirtualMachinePublishRequestContext) error {
+	vmPubReq := ctx.VMPublishRequest
+	// If this annotation is not present on the VM Publish Request, then we have not begun quota
+	// validation. We need to apply this annotation along with the requested capacity annotation
+	// and await processing.
+	if _, ok := vmPubReq.Annotations[pkgconst.AsyncQuotaPerformCheckAnnotationKey]; !ok {
+		vm := ctx.VM
+		if vm.Status.Storage != nil && vm.Status.Storage.Total != nil {
+			if vmPubReq.Annotations == nil {
+				vmPubReq.Annotations = make(map[string]string)
+			}
+			vmPubReq.Annotations[pkgconst.AsyncQuotaPerformCheckAnnotationKey] = "true"
+			vmPubReq.Annotations[pkgconst.AsyncQuotaCheckRequestedCapacityAnnotationKey] = vm.Status.Storage.Total.String()
+
+			return fmt.Errorf("quota validation is needed for this request")
+		}
+		return fmt.Errorf("unable get storage total for VM %q for quota validation", vm.Name)
+	}
+	// Check for the existence of the requested-quota annotation. If this annotation is present, then we are
+	// currently awaiting quota verification, and we should halt any further processing.
+	if _, ok := vmPubReq.Annotations[pkgconst.AsyncQuotaCheckRequestedCapacityAnnotationKey]; ok {
+
+		return fmt.Errorf("quota validation has not completed for this request")
```

Comment:
same as above. not sure if we need to return an error for re-queues since annotation change will trigger reconciliation anyway